images/router/egress/egress-router.sh

```diff
@@ -1,57 +1,67 @@
-    ip addr add ${EGRESS_SOURCE}/32 dev macvlan0
+if ! ip addr show dev macvlan0 | grep -q "${EGRESS_SOURCE}"; then
+    ip addr add "${EGRESS_SOURCE}"/32 dev macvlan0
     ip link set up dev macvlan0
-    ip route add ${EGRESS_GATEWAY}/32 dev macvlan0
-    if [ "${EGRESS_DESTINATION}" != "${EGRESS_GATEWAY}" ]; then
-        ip route add "${EGRESS_DESTINATION}"/32 via "${EGRESS_GATEWAY}" dev macvlan0
-    fi
-    iptables -t nat -A PREROUTING -i eth0 -j DNAT --to-destination ${EGRESS_DESTINATION}
-    iptables -t nat -A POSTROUTING -j SNAT --to-source ${EGRESS_SOURCE}
+
+    ip route add "${EGRESS_GATEWAY}"/32 dev macvlan0
+    ip route del default
+    ip route add default via "${EGRESS_GATEWAY}" dev macvlan0
 fi
 
 # Update neighbor ARP caches in case another node previously had the IP. (This is
 # the same code ifup uses.)
-arping -q -A -c 1 -I macvlan0 ${EGRESS_SOURCE}
+arping -q -A -c 1 -I macvlan0 "${EGRESS_SOURCE}"
 ( sleep 2;
-  arping -q -U -c 1 -I macvlan0 ${EGRESS_SOURCE} || true ) > /dev/null 2>&1 < /dev/null &
-
-# Now we just wait until we are killed...
-#
-# Kubernetes will use SIGTERM to kill us, but bash ignores SIGTERM by
-# default in interactive shells, and it thinks this shell is
-# interactive due to the way in which docker invokes it. We can get
-# bash to react to SIGTERM if we explicitly set a trap for it, except
-# that bash doesn't process signal traps while it is waiting for a
-# process to finish, and we have to be waiting for a process to finish
-# because there's no way to sleep forever within bash.
-#
-# Fortunately, signal traps do interrupt the "wait" builtin. So...
-# set up a SIGTERM trap, run a command that sleeps forever *in the
-# background*, and then wait for either the command to finish or the
-# signal to arrive.
-
-trap "exit" TERM
-tail -f /dev/null &
-wait
-
-# We don't have to do any cleanup because deleting the network
-# namespace will clean everything up for us.
+  arping -q -U -c 1 -I macvlan0 "${EGRESS_SOURCE}" || true ) > /dev/null 2>&1 < /dev/null &
+
+if [[ -z "${EGRESS_DESTINATION:-}" ]]; then
+    ### HTTP Proxy Mode
+    exec squid -N
+
+else
+    ### IPTables Redirection Mode
+
+    if ! iptables -t nat -C PREROUTING -i eth0 -j DNAT --to-destination "${EGRESS_DESTINATION}"; then
+	iptables -t nat -A PREROUTING -i eth0 -j DNAT --to-destination "${EGRESS_DESTINATION}"
+	iptables -t nat -A POSTROUTING -j SNAT --to-source "${EGRESS_SOURCE}"
+    fi
+
+    # Now we just wait until we are killed...
+    #
+    # Kubernetes will use SIGTERM to kill us, but bash ignores SIGTERM by
+    # default in interactive shells, and it thinks this shell is
+    # interactive due to the way in which docker invokes it. We can get
+    # bash to react to SIGTERM if we explicitly set a trap for it, except
+    # that bash doesn't process signal traps while it is waiting for a
+    # process to finish, and we have to be waiting for a process to finish
+    # because there's no way to sleep forever within bash.
```

Comment:
It doesn't leave a zombie (note that this is existing code, it's just getting re-indented). But yeah, a loop-and-sleep would be more obvious anyway...

I'm also thinking about making the setup be done in an initContainer, and then run origin-pod as the "main" container. (Or is that "cheating" because in the future origin-pod might do something other than just sit and wait for a SIGTERM?)